commit 1235441af59c89c9989de236d2d816d1555d08b1
Author: Peter Jones <pjones@redhat.com>
Date:   Mon Nov 3 10:38:43 2014 -0500

    Correctly return error and report errno on read failures.
    
    Based on the bug and patch reported in this thread:
    https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=764386
    
    Signed-off-by: Peter Jones <pjones@redhat.com>

diff --git a/src/util.h b/src/util.h
index 94708e1..fdeb649 100644
--- a/src/util.h
+++ b/src/util.h
@@ -45,10 +45,12 @@ read_file(int fd, uint8_t **buf, size_t *bufsize)
 		if (s < 0 && errno == EAGAIN) {
 			continue;
 		} else if (s < 0) {
+			int saved_errno = errno;
 			free(*buf);
 			*buf = NULL;
 			*bufsize = 0;
-			break;
+			errno = saved_errno;
+			return -1;
 		}
 		filesize += s;
 		/* only exit for empty reads */
@@ -57,9 +59,11 @@ read_file(int fd, uint8_t **buf, size_t *bufsize)
 		} else if (s == 4096) {
 			newbuf = realloc(*buf, size + 4096);
 			if (newbuf == NULL) {
+				int saved_errno = errno;
 				free(*buf);
 				*buf = NULL;
 				*bufsize = 0;
+				errno = saved_errno;
 				return -1;
 			}
 			*buf = newbuf;
